vpr/src/device/rr_spatial_lookup.cpp

```diff
@@ -0,0 +1,93 @@
+#include "vtr_assert.h"
+#include "rr_spatial_lookup.h"
+
+RRSpatialLookup::RRSpatialLookup(t_rr_node_indices& rr_node_indices)
+    : rr_node_indices_(rr_node_indices) {
+}
+
+RRNodeId RRSpatialLookup::find_node(int x,
+                                    int y,
+                                    t_rr_type type,
+                                    int ptc,
+                                    e_side side) const {
+    /* Find actual side to be used */
+    e_side node_side = side;
+    if (type == IPIN || type == OPIN) {
+        VTR_ASSERT_MSG(side != NUM_SIDES, "IPIN/OPIN must specify desired side (can not be default NUM_SIDES)");
+    } else {
+        VTR_ASSERT(type != IPIN && type != OPIN);
+        node_side = SIDES[0];
+    }
+
+    /* Currently need to swap x and y for CHANX because of chan, seg convention 
+     * This is due to that the fast look-up builders uses (y, x) coordinate when
+     * registering a CHANX node in the look-up
+     * TODO: Once the builders is reworked for use consistent (x, y) convention,
+     * the following swapping can be removed
+     */
+    size_t node_x = x;
+    size_t node_y = y;
+    if (CHANX == type) {
+        std::swap(node_x, node_y);
+    }
+
+    VTR_ASSERT_SAFE(3 == rr_node_indices_[type].ndims());
+
+    /* Sanity check to ensure the x, y, side and ptc are in range 
+     * - Return an valid id by searching in look-up when all the parameters are in range
+     * - Return an invalid id if any out-of-range is detected
+     */
+    if (size_t(type) >= rr_node_indices_.size()) {
+        return RRNodeId::INVALID();
+    }
+
+    if (node_x >= rr_node_indices_[type].dim_size(0)) {
+        return RRNodeId::INVALID();
+    }
+
+    if (node_y >= rr_node_indices_[type].dim_size(1)) {
+        return RRNodeId::INVALID();
+    }
+
+    if (node_side >= rr_node_indices_[type].dim_size(2)) {
+        return RRNodeId::INVALID();
+    }
```

Comment:
Agree. However, the ``node_x``, ``node_y`` are already in the ``size_t`` data type, they cannot be less than 0 (converted from ``x ``, ``y`` which are ``int``). If ``x`` is negative, when converted to ``size_t``, it will be a huge number and out-of-range in the fast look-up. But it is really tricky. So, I added a pre-check on the ``x``,  ``y``, ``node_side`` and ``ptc`` to return invalid ids when one of them is negative.

Reply:
Ah, I didn't think of that.  In that case I'm OK with just a comment noting that will happen.  If you've added the pre-check for < 0 already that's fine too.